circuits/plonk-plookup/src/gate.rs

```diff
@@ -0,0 +1,123 @@
+/*****************************************************************************************************************
+
+This source file implements Plonk constraint gate primitive.
+
+*****************************************************************************************************************/
+
+use algebra::FftField;
+pub use super::{wires::{*}, constraints::ConstraintSystem};
+use std::io::{Read, Result as IoResult, Write, Error, ErrorKind};
+use num_traits::cast::{FromPrimitive, ToPrimitive};
+use algebra::bytes::{FromBytes, ToBytes};
+
+#[repr(C)]
+#[derive(Clone, Debug)]
+#[derive(PartialEq)]
+#[derive(FromPrimitive, ToPrimitive)]
+pub enum GateType
+{
+    Zero,       // zero gate
+    Generic,    // generic arithmetic gate
+    Poseidon,   // Poseidon permutation gate
+    Add,        // EC addition in Affine form
+    Double,     // EC point doubling in Affine form
+    Vbmul1,     // EC variable base scalar multiplication 
+    Vbmul2,     // unpacking EC variable base scalar multiplication 
+    Endomul,    // EC variable base scalar multiplication with group endomorphim optimization
+    Pack,       // packing
+    Lookup,     // lookup
+}
+
+#[derive(Clone)]
+pub struct CircuitGate<F: FftField>
```

Comment:
Let's reuse this and the ToBytes/FromBytes impls with non-plookup plonk. I think a good way to do it would be to parametrize `CircuitGate` over `<Gt : GateType>` where `GateType` is a trait

```
trait GateType {
     fn verify(&self, witness: &[Vec<F>; COLUMNS], cs: &ConstraintSystem<F>) -> bool 
}
```

and then in the implementation for this trait, you just match on the `self` parameter and do the appropriate thing.

Then, for plookup plonk, we can have the GateType be an enum which is either a GateType value from regular plonk, or a lookup gate. That we, we can also easily reuse all the verification code with PLONK.